share/completions/zpool.fish

```diff
@@ -37,20 +37,20 @@ function __fish_zpool_using_command # zpool command whose completions are looked
 end
 
 function __fish_zpool_list_used_vdevs -a pool
-    if test -z "$pool"
-        zpool list -H -v | grep -E "^\s" | cut -f2 | grep -x -E -v "(spare|log|cache|mirror|raidz.?)"
-    else
-        zpool list -H -v $pool | grep -E "^\s" | cut -f2 | grep -x -E -v "(spare|log|cache|mirror|raidz.?)"
-    end
+    zpool list -H -v $pool | string replace -rf '^\s' '' | string match -rf '^(spare|log|cache|mirror|raidz.?)'
```

Comment:
There's no need for the `string` in the middle at all, as `zpool -H` is intended exactly for uses like this (for use in scripts) and is documented as separating fields by a single `\t` instead of an arbitrary number of spaces. This is better because a) we don't need a separate command to collapse whitespace, b) we won't ever confuse spaces that are part of a field with field separators.

In all cases, neither this nor the old version work under FreeBSD. `cut -f2` returns the size of each pool rather than the vdevs in it.

```
mqudsi@freebsd /u/h/m/fish-shell> zpool list -H -v
tank    3.62T   2.40T   1.23T   -       -       18%     66%     1.00x   ONLINE  -
        mirror  3.62T   2.40T   1.23T   -       -       18%     66%
        da1     -       -       -       -       -       -       -
        da2     -       -       -       -       -       -       -
cache                                       -      -      -         -      -      -
        nvd0    477G    374M    477G    -       -       0%      0%
zroot   236G    40.4G   196G    -       -       18%     17%     1.00x   ONLINE  -
        gptid/c3ae7a7e-bc4c-11e3-ab8c-4437e6acfe08      236G    40.4G   196G    -       -       18%     17%
```

becomes

```
mqudsi@freebsd /u/h/m/fish-shell> zpool list -H -v | string replace -rf '^\s' '' | cut -f2
3.62T
-
-
477G
236G
```

`cut -f1` is similarly confused because the `vdev` lines don't have a first column:

```
mqudsi@freebsd /u/h/m/fish-shell> zpool list -Hv | string replace -r '\s+' ' ' | cut -f1
tank 3.62T
 mirror
 da1
 da2
cache -      -      -         -      -      -
 nvd0
zroot 236G
 gptid/c3ae7a7e-bc4c-11e3-ab8c-4437e6acfe08
```

I think this is what you want (which means that incorrect `string match -rf` was supposed to be `string match -rv`):


```
mqudsi@freebsd /u/h/m/fish-shell> zpool list -Hv | string match -vr '^\S' | cut -d\t -f2 | string match -rv '
^(spare|log|cache|mirror|raidz.?)'
da1
da2
nvd0
gptid/c3ae7a7e-bc4c-11e3-ab8c-4437e6acfe08
```

Reply:
*I also think this probably is the same for other platforms.*

Reply:
>There's no need for the string in the middle at all, as zpool -H is intended exactly for uses like this (for use in scripts) and is documented as separating fields by a single \t instead of an arbitrary number of spaces. 

The "string in the middle" does two things:

- It matches only lines starting with whitespace
- It removes that whitespace

Both of which are necessary here, to exclude the header lines and not offer whitespace as the completion.

>In all cases, neither this nor the old version work under FreeBSD. cut -f2 returns the size of each pool rather than the vdevs in it.

No, the `cut -f2` was used with `grep`, which only matched but didn't do the removal. The `string replace` replaces `grep | cut` both.

>which means that incorrect string match -rf was supposed to be string match -rv

It was indeed. Good catch!

Reply:
The new replacement still doesn't work:

```
$ zpool list -H -v $pool | string replace -rf '^\s' '' | string match -rv '^(spare|log|cache|mirror|raidz.?)'
da1	-	-	-	-	-	-	-
da2	-	-	-	-	-	-	-
nvd0	477G	389M	477G	-	-	0%	0%
gptid/c3ae7a7e-bc4c-11e3-ab8c-4437e6acfe08	236G	41.0G	195G	-	-	18%	17%
```

I replaced it with this variant of my proposal instead, swapping out the `cut` for `string` and then merging two of the `string` invocations together:

```
zpool list -Hv | string replace -rf "^\t([^\t]+).*" '$1' | string match -rv '^(spare|log|cache|mirror|raidz.?)'
```